database/database.go

```diff
@@ -153,6 +153,93 @@ func (db DatastoreBacked) StateForDomain(domain string) (state DomainState, err
 	return state, nil
 }
 
+// StatesForDomains returns the domains states for the given domains
+func (db DatastoreBacked) StatesForDomains(domains []string) (states []DomainState, err error) {
+	// Set up datastore context
+	c, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	client, datastoreErr := db.backend.NewClient(c, db.projectID)
+	if datastoreErr != nil {
+		return states, datastoreErr
+	}
+
+	getDomainStates := func(keys []*datastore.Key) ([]DomainState, error) {
+		domainStates := make([]DomainState, len(keys))
+		if err := client.GetMulti(c, keys, domainStates); err != nil {
+			return nil, err
+		}
+		for i := range domainStates {
+			domainStates[i].Name = keys[i].Name
+		}
+		return domainStates, nil
+	}
+
+	var keys []*datastore.Key
+	for _, domain := range domains {
+		key := datastore.NameKey(domainStateKind, domain, nil)
+		keys = append(keys, key)
+		if len(keys) >= batchSize {
```

Comment:
I'm not sure I follow this section, this is getting the domain states for a subset of keys (< batchsize), but then not doing anything with them (unless there is an error), then at the end what we actually return is getDomainStates for all the keys. A similar approach works in the put case because each call results in an update to the DB, but since we are just reading in this case, I think this code basically is a no-op (unless I'm missing something).

Reply:
I left out a part in this section. If len(keys) exceeded batch size, then I was going to store the output of the call to getDomainStates in a slice. Then, when we call getDomainStates on the remaining keys at the end of the function, I was going append this output to the slice from earlier. Do you think that this is neccessary?

Reply:
Done

Reply:
Just updated the PR with the change I proposed